doc/user: improve Kafka source and sink docs (#20196)
MaterializeInc/materialize: doc/user/content/sql/create-sink/kafka.md

`SNAPSHOT`           | `bool` | Default: `true`. Whether to emit the consolidated results of the query before the sink was created at the start of the sink. To see only results after the sink is created, specify `WITH (SNAPSHOT = false)`.
`SIZE`               | `text`    | The [size](#sizing-a-sink) for the sink. Accepts values: `3xsmall`, `2xsmall`, `xsmall`, `small`, `medium`, `large`, `xlarge`. Required if the `IN CLUSTER` option is not specified.

## Formats

The `FORMAT` option controls the encoding of the message key and value that
Materialize writes to Kafka.

**Known limitation:** The key format cannot differ from the value format.

### Avro

<b>Syntax:</b> `FORMAT JSON`

The Avro format encodes each Kafka message value as an Avro record containing a
field for each column of the sink's source relation. The names and ordering of
the fields in the record match the names and ordering of the columns in the

> relation. SQL types are converted to Avro types according to the type conversion

Author: Done.

table below.

> If the `KEY` option is specified, the Avro format encodes each Kafka message key

Author: Ah, buried in the paragraph is "Kafka message value" vs "Kafka message key". The `KEY` options additional key metadata to each message. I've reworded to try to make that more clear. What do you think of the new phrasing? I still don't love it, but I think it's better.

according to the same rules, except the encoded Avro record will contain only
the specified key columns.

When using a Confluent Schema Registry, you can specify the
[fullnames](https://avro.apache.org/docs/current/specification/#names) for the
Avro schemas Materialize generates using the `AVRO KEY FULLNAME` and `AVRO VALUE
FULLNAME` [syntax](#syntax).

<br>
<details>

> <summary>Type conversion table</summary>

Author: Done.

[`numeric(p,s)`][`numeric`]  | `{"type": "bytes", "logicalType": "decimal", "precision": P, "scale": s}`
[`oid`]                      | `{"type": "fixed", "size": 4, "name": "com.materialize.sink.uint4"}`
[`real`]                     | `"float"`
[`record`]                   | `{"type": "record", "name": ..., "fields": ...}`
[`smallint`]                 | `"int"`
[`text`]                     | `"string"`
[`time`]                     | `{"type": "long", "logicalType": "time-micros"}`
[`uint2`]                    | `{"type": "fixed", "size": 2, "name": "com.materialize.sink.uint2"}`
[`uint4`]                    | `{"type": "fixed", "size": 4, "name": "com.materialize.sink.uint4"}`
[`uint8`]                    | `{"type": "fixed", "size": 8, "name": "com.materialize.sink.uint8"}`
[`timestamp`]                | `{"type": "long", "logicalType: "timestamp-micros"}`
[`timestamp with time zone`] | `{"type": "long", "logicalType: "timestamp-micros"}`
[Arrays]                     | `{"type": "array", "items": ...}`
</details>

### JSON

<b>Syntax:</b> `FORMAT JSON`

The JSON format encodes each Kafka message value as a JSON object containing a

> field for each column of the sinked relation. The names and ordering of the

Author: Oh good catch, thanks!

fields in the record match the names and ordering of the columns in the
relation.

> If the `KEY` option is specified, the JSON format encodes each Kafka message key

Author: (Same response as for Avro.)

are known to be unique for the underlying relation:
  ("col2")
  ("col3", "col4")
```

The first error message indicates that Materialize could not prove the existence
of any unique keys for the sink's source relation. The second error message
indicates that Materialize could prove that `col2` and `(col3, col4)` were
unique keys of the sink's source relation, but could not provide the uniqueness
of the specified upsert key of `col1`.

There are three ways to resolve this error:

* Change the sink to use one of the keys that Materialize determined to be
  unique, if such a key exists and has the appropriate semantics for your
  use case.

* Create a materialized view that deduplicates the input relation by the
  desired upsert key:

  ```sql
  -- For each row with the same key `k`, the `ORDER BY` clause ensures we
  -- keep the row with the largest value of `v`.
  CREATE MATERIALIZED VIEW deduped AS
  SELECT DISTINCT ON (k) v
  FROM original_input
  ORDER BY k, v DESC;

  -- Materialize can now prove that `k` is a unique key of `deduped`.
  CREATE SINK s
  FROM deduped
  INTO KAFKA CONNECTION kafka_connection (TOPIC 't')
  KEY (k)

>   FORMAT JSON ENVELOPE UPSERT WITH (SIZE = '2xmsall');

Author: Done.

  number of records in `original_input`. Be sure to assign `deduped`
  to a cluster with resources sufficient to handle your data volume.
  {{< /note >}}

* Use the `NOT ENFORCED` clause to disable Materialize's validation of the key's
  uniqueness:

  ```sql
  CREATE SINK s
  FROM original_input
  INTO KAFKA CONNECTION kafka_connection (TOPIC 't')
  -- We have outside knowledge that `k` is a unique key of `original_input`, but
  -- Materialize cannot prove this, so we disable its key uniqueness check.
  KEY k NOT ENFORCED
  FORMAT JSON ENVELOPE UPSERT WITH (SIZE = '2xsmall');
  ```

  You should only disable this verification if you have outside knowledge of
  the properties of your data that guarantees the uniqueness of the key you
  have specified.

  {{< warning >}}
  If the key is not in fact unique, downstream consumers may not be able to
  correctly interpret the data in the topic, and Kafka key compaction may
  incorrectly garbage collect records from the topic.
  {{< /warning >}}

### Debezium

<b>Syntax:</b> `ENVELOPE DEBEZIUM`

The Debezium envelope wraps each event in an object containing a `before` and
`after` field:

> ```json

Author: Done.

> ```json

Author: Yeah, I wanted the JSON highlighting! GitHub can't handle the `//` comments and the `...` but our docs handle it nicely.

`SNAPSHOT`           | `bool` | Default: `true`. Whether to emit the consolidated results of the query before the sink was created at the start of the sink. To see only results after the sink is created, specify `WITH (SNAPSHOT = false)`.
`SIZE`               | `text`    | The [size](#sizing-a-sink) for the sink. Accepts values: `3xsmall`, `2xsmall`, `xsmall`, `small`, `medium`, `large`, `xlarge`. Required if the `IN CLUSTER` option is not specified.

## Formats

The `FORMAT` option controls the encoding of the message key and value that
Materialize writes to Kafka.

**Known limitation:** The key format cannot differ from the value format.

### Avro

> <b>Syntax:</b> `FORMAT JSON`

Reviewer: ```suggestion
<b>Syntax:</b> `FORMAT AVRO`
```

Author: Ugh, thanks, fixed!

  INTO KAFKA CONNECTION kafka_connection (TOPIC 't')
  -- We have outside knowledge that `k` is a unique key of `original_input`, but
  -- Materialize cannot prove this, so we disable its key uniqueness check.
  KEY k NOT ENFORCED
  FORMAT JSON ENVELOPE UPSERT;
  ```

  You should only disable this verification if you have outside knowledge of
  the properties of your data that guarantees the uniqueness of the key you
  have specified.

  {{< warning >}}
  If the key is not in fact unique, downstream consumers may not be able to
  correctly interpret the data in the topic, and Kafka key compaction may
  incorrectly garbage collect records from the topic.
  {{< /warning >}}

### Debezium

<b>Syntax:</b> `ENVELOPE DEBEZIUM`

The Debezium envelope wraps each event in an object containing a `before` and
`after` field to indicate whether the event was an insertion, deletion, or
update:

```json
// Insertion event.
{"before": null, "after": {"field1": "val1", ...}}

// Deletion event.
{"before": {"field1": "val1", ...}, "after": null}

// Update event.
{"before": {"field1": "oldval1", ...}, "after": {"field1": "newval1", ...}}
```

Consider using the Debezium envelope if:

>   * You have downstream consumers that want update events to contain both the

Reviewer 2: nit weird to have a single-element bulleted list

Author: Oh good point. I missed the second bullet that I meant to add here ("there is no natural key for the sink").

[`text`]                     | `"string"`
[`time`]                     | `{"type": "long", "logicalType": "time-micros"}`
[`uint2`]                    | `{"type": "fixed", "size": 2, "name": "com.materialize.sink.uint2"}`
[`uint4`]                    | `{"type": "fixed", "size": 4, "name": "com.materialize.sink.uint4"}`
[`uint8`]                    | `{"type": "fixed", "size": 8, "name": "com.materialize.sink.uint8"}`
[`timestamp`]                | `{"type": "long", "logicalType: "timestamp-micros"}`
[`timestamp with time zone`] | `{"type": "long", "logicalType: "timestamp-micros"}`
[Arrays]                     | `{"type": "array", "items": ...}`

### JSON

<b>Syntax:</b> `FORMAT JSON`

When using the Avro format, the value of each Kafka message is a JSON object
containing a field for each column of the sink's source relation. The names and
ordering of the fields in the record match the names and ordering of the columns
in the relation.

If the `KEY` option is specified, the key of each Kafka message is a JSON
object containing a field for each key column, in the same order and with the
same names.

SQL values are converted to JSON values according to the following rules:

>   * [`boolean`] values are converted to `true` or `false`.
>   * [`smallint`], [`integer`], [`bigint`], [`uint2`], [`uint4`], [`uint8`],
>     [`real`], and [`double precision`] values are converted to JSON numbers.
>   * [`numeric`] values are converted to a JSON string containing the
>     decimal representation of the number.
>   * [`record`] types are converted to JSON objects. The names and ordering of
>     the fields in the object match the names and ordering of the fields in the
>     record.
>   * [`list`] values are converted to JSON arrays.
>   * [array][arrays] values are converted to JSON arrays. Multidimensional arrays
>     are flattened into a single dimension following row-major order.
>   * [`timestamp`] and [`timestamp with time zone`] values are converted to JSON
>     strings containing the number of milliseconds since the Unix epoch.
>   * All other types are cast to [`text`] and then converted to JSON strings.

Reviewer 3: ```suggestion
SQL type                     | Conversion
-----------------------------|-------------------------------------
[`array`][`arrays`]          | Values are converted to JSON arrays. Multidimensional arrays are flattened into a single dimension following row-major order.
[`bigint`]                   | Values are converted to JSON numbers.
[`boolean`]                  | Values are converted to `true` or `false`.
[`integer`]                  | Values are converted to JSON numbers.
[`list`]                     | Values are converted to JSON arrays.
[`numeric`]                  | Values are converted to a JSON string containing the decimal representation of the number.
[`record`]                   | Records are converted to JSON objects. The names and ordering of the fields in the object match the names and ordering of the fields in the record.
[`smallint`]                 | values are converted to JSON numbers.
[`timestamp`][`timestamp with time zone`] | Values are converted to JSON strings containing the number of milliseconds since the Unix epoch.
[`uint2`]                    | Values are converted to JSON numbers.
[`uint4`]                    | Values are converted to JSON numbers.
[`uint8`]                    | Values are converted to JSON numbers.
Other                        | Values are cast to [`text`] and then converted to JSON strings.
```

## Envelopes

> The sink's envelope determines how inserts, updates, and delete events are

Reviewer 3: ```suggestion
In addition to determining how to encode output data, Materialize also needs to
understand the desired behaviour of sinked events in the downstream external
system. Whether a change is simply emitted as an event with a diff structure,
or actively inserts, updates, or deletes existing data downstream depends on
the `ENVELOPE` specified in the `CREATE SINK` statement.
```

mapped to Kafka messages.

### Upsert

> <b>Syntax:</b> `ENVELOPE UPSERT`

Reviewer 3: ```suggestion
<p style="font-size:14px"><b>Syntax:</b> <code>ENVELOPE UPSERT</code></p>
```

> The upsert envelope:

Reviewer 3: ```suggestion
The upsert envelope treats all records as having a **key** and a **value**, and
supports inserts, updates and deletes in the sink:
```

Author: Similar thought to the above. It's hard for me to square "supporting updates and deletes" with what the upsert envelope actually does.

  * Emits insertion and update events without additional decoration.

>   * Converts deletion events into messages with a `null` value (often called
>     "tombstones").

Reviewer 3: ```suggestion
  * Converts deletion events into messages with a `NULL` value (i.e. _tombstones_).
```

  * Requires that you specify a unique key for the sink's source
    relation using the `KEY` option.

Consider using the upsert envelope if:

  * You need to follow standard Kafka conventions for upsert semantics.
  * You want to enable key-based compaction on the sink's Kafka topic while
    retaining the most recent value for each key.

> #### Key selection

Reviewer 3: The content in this section is very troubleshooting-focused, so we're again mixing and matching reference documentation and guides, but we can figure out how to move content around later.

indicates that Materialize could prove that `col2` and `(col3, col4)` were
unique keys of the sink's source relation, but could not provide the uniqueness
of the specified upsert key of `col1`.

There are three ways to resolve this error:

* Change the sink to use one of the keys that Materialize determined to be
  unique, if such a key exists and has the appropriate semantics for your
  use case.

* Create a materialized view that deduplicates the input relation by the
  desired upsert key:

  ```sql
  -- For each row with the same key `k`, the `ORDER BY` clause ensures we
  -- keep the row with the largest value of `v`.
  CREATE MATERIALIZED VIEW deduped AS
  SELECT DISTINCT ON (k) v
  FROM original_input
  ORDER BY k, v DESC;

  -- Materialize can now prove that `k` is a unique key of `deduped`.
  CREATE SINK s IN CLUSTER my_io_cluster
  FROM deduped
  INTO KAFKA CONNECTION kafka_connection (TOPIC 't')
  KEY (k)
  FORMAT JSON ENVELOPE UPSERT;
  ```

  {{< note >}}
  Maintaining the `deduped` materialized view requires memory proportional to the
  number of records in `original_input`. Be sure to assign `deduped`

>   to a cluster with resources sufficient to handle your data volume.

Reviewer 3: ```suggestion
  to a cluster with adequate resources to handle your data volume.
```

  {{< /note >}}

* Use the `NOT ENFORCED` clause to disable Materialize's validation of the key's
  uniqueness:

  ```sql
  CREATE SINK s IN CLUSTER my_io_cluster
  FROM original_input
  INTO KAFKA CONNECTION kafka_connection (TOPIC 't')
  -- We have outside knowledge that `k` is a unique key of `original_input`, but
  -- Materialize cannot prove this, so we disable its key uniqueness check.
  KEY k NOT ENFORCED
  FORMAT JSON ENVELOPE UPSERT;
  ```

  You should only disable this verification if you have outside knowledge of
  the properties of your data that guarantees the uniqueness of the key you
  have specified.

  {{< warning >}}
  If the key is not in fact unique, downstream consumers may not be able to
  correctly interpret the data in the topic, and Kafka key compaction may
  incorrectly garbage collect records from the topic.
  {{< /warning >}}

### Debezium

> <b>Syntax:</b> `ENVELOPE DEBEZIUM`

Reviewer 3: ```suggestion
<p style="font-size:14px"><b>Syntax:</b> <code>ENVELOPE DEBEZIUM</code></p>
```

> The Debezium envelope wraps each event in an object containing a `before` and
> `after` field to indicate whether the event was an insertion, deletion, or
> update:

Reviewer 3: ```suggestion
The Debezium envelope wraps each event in an object containing a `before` and
`after` field to indicate whether the event was an insertion, deletion, or
update to the underlying source, table or materialized view being written to
the sink:
```

For more details, see [the Kafka documentation](https://kafka.apache.org/documentation/).

## Required permissions

> The access control lists (ACLs) on the Kafka cluster must permit Materialize

Reviewer 3: ```suggestion
The access control lists (ACLs) on the Kafka cluster must allow Materialize
```

to perform the following operations on the following resources:

> Operations  | Resource type    | Resource name

Reviewer 3: ```suggestion
Operation type | Resource type    | Resource name
```

------------|------------------|--------------
Read, Write | Topic            | Consult `mz_kafka_connections.sink_progress_topic` for the sink's connection
Write       | Topic            | The specified `TOPIC` option
Write       | Transactional ID | `mz-producer-{SINK ID}-*`

When using [automatic topic creation](#automatic-topic-creation), Materialize
additionally requires access to the following operations:

> Operations       | Resource type    | Resource name

Reviewer 3: ```suggestion
Operation type       | Resource type    | Resource name
```

### Creating a sink

#### Upsert envelope

{{< tabs tabID="1" >}}
{{< tab "Avro">}}

```sql
CREATE SINK avro_sink
  FROM <source, table or mview>
  INTO KAFKA CONNECTION kafka_connection (TOPIC 'test_avro_topic')
  KEY (key_col)
  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_connection
  ENVELOPE DEBEZIUM
  WITH (SIZE = '3xsmall');
```

{{< /tab >}}
{{< tab "JSON">}}

```sql
CREATE SINK json_sink
  FROM <source, table or mview>
  INTO KAFKA CONNECTION kafka_connection (TOPIC 'test_json_topic')
  KEY (key_col)
  FORMAT JSON
  ENVELOPE DEBEZIUM

>   WITH (SIZE = '3xsmall');

Reviewer 4: Same comment as before. If we want to start discouraging linked clusters, we might want to use `IN CLUSTER` here.

Author: Done.

```

{{< /tab >}}
{{< /tabs >}}

#### Debezium envelope

> {{< tabs tabID="1" >}}

Reviewer 4: I don't *think* you need this?

Author: Good point. Removed.
